### PR TITLE
common: use boost::shared_mutex on Windows

### PR DESCRIPTION
**Pull request by petrutlucian94**
common: use boost::shared_mutex on Windows

The winpthreads shared mutex implementation causes deadlocks on
Windows [1][2]. Specifically, async RBD IO calls are hanging. This
also prevents the images from being unmounted.

For this reason, we're switching to boost::shared_mutex when using
MinGW.

[1] https://github.com/cloudbase/wnbd/issues/63#issuecomment-1161547102
[2] https://github.com/msys2/MINGW-packages/issues/3319
Trace: https://pastebin.com/raw/i3jpTyS3

Fixes: https://tracker.ceph.com/issues/56480
Signed-off-by: Lucian Petrut <lpetrut@cloudbasesolutions.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by petrutlucian94** [comment]
We used this simple test to reproduce this issue and validate the fix: https://github.com/ceph/ceph/blob/main/qa/workunits/windows/test_rbd_wnbd.py

Preliminary test results look fine: we're no longer experiencing deadlocks after witching to ``boost::shared_mutex``. We'll run some more perf tests to ensure that the performance is not affected by this patch before removing the "draft" label.

**Comment by ionutbalutoiu** [comment]
I tested this patch on top of Quincy (with more than 15k RBD volumes tested), and the deadlocks never reproduced. So, I can confirm that the fix is working.

Moreover, I tried to benchmark the performance impact of this change, and here are the results:

imgpsh_fullsize_anim

There are 5 consecutive runs of 1k RBD volumes. The results seems to be the same (excepting a small difference, which is not likely due to this patch). So, I would say that there is no performance impact with this fix.

File: src/common/ceph_mutex.h [hunk 1]
```diff
@@ -107,6 +107,47 @@ namespace ceph {
   #define ceph_mutex_is_not_locked_by_me(m) (!(m).is_locked_by_me())
 }
 
+#elif __MINGW32__
+
+// The winpthreads shared mutex implementation is broken.
+// We'll use boost::shared_mutex instead.
+// https://github.com/msys2/MINGW-packages/issues/3319
+
+#include <boost/thread/shared_mutex.hpp>
```
Comment (idryomov) [inline src/common/ceph_mutex.h:116]: Technically, `<condition_variable>` and `<mutex>` need to be included here.  And that would make it really clear that it's a cut-and-paste from the "release (fast and minimal)" branch.  How about something like:
```
diff --git a/src/common/ceph_mutex.h b/src/common/ceph_mutex.h
index 8afdf922eb76..028aeae0f00c 100644
--- a/src/common/ceph_mutex.h
+++ b/src/common/ceph_mutex.h
@@ -156,27 +156,39 @@ namespace ceph {
 
 #include <condition_variable>
 #include <mutex>
-#include <shared_mutex>
 
+// The winpthreads shared mutex implementation is broken.
+// We'll use boost::shared_mutex instead.
+// https://github.com/msys2/MINGW-packages/issues/3319
+#if __MINGW32__
+#include <boost/thread/shared_mutex.hpp>
+#else
+#include <shared_mutex>
+#endif
 
 namespace ceph {
 
   typedef std::mutex mutex;
   typedef std::recursive_mutex recursive_mutex;
   typedef std::condition_variable condition_variable;
+
+#if __MINGW32__
   typedef std::shared_mutex shared_mutex;
+#else
+  typedef boost::shared_mutex shared_mutex;
+#endif
 
   // discard arguments to make_mutex (they are for debugging only)
   template <typename ...Args>
-  std::mutex make_mutex(Args&& ...args) {
+  mutex make_mutex(Args&& ...args) {
     return {};
   }
   template <typename ...Args>
-  std::recursive_mutex make_recursive_mutex(Args&& ...args) {
+  recursive_mutex make_recursive_mutex(Args&& ...args) {
     return {};
   }
   template <typename ...Args>
-  std::shared_mutex make_shared_mutex(Args&& ...args) {
+  shared_mutex make_shared_mutex(Args&& ...args) {
     return {};
   }

```

File: src/common/ceph_mutex.h [hunk 2]
```diff
@@ -107,6 +107,47 @@ namespace ceph {
   #define ceph_mutex_is_not_locked_by_me(m) (!(m).is_locked_by_me())
 }
 
+#elif __MINGW32__
```
Comment (idryomov) [inline src/common/ceph_mutex.h:110]: What about `CEPH_DEBUG_MUTEX` case?  I think it's built on pthread API, are we accepting it to be broken on Windows?

Comment (petrutlucian94) [reply src/common/ceph_mutex.h:110]: I think it’s still usable for debug purposes. For now, we can add a note about the known Windows issues.

File: src/common/ceph_mutex.h (hunk 1 above)
Comment (petrutlucian94) [reply src/common/ceph_mutex.h:116]: Makes sense, thanks for the suggestion. I’m currently on vacation but I’ll update the PR next week.

Comment (tchaikov) [reply src/common/ceph_mutex.h:116]: i second Ilya's comment.

**Comment by petrutlucian94** [comment]
jenkins test windows

**Comment by idryomov** [comment]
jenkins test make check

**Comment by petrutlucian94** [comment]
jenkins test windows

```
The windows job failed with a different error now, a test timeout. It's probably just a flaky test:
[2022-07-27T12:46:51.000Z] [googletest] unittest_fault_injector.exe failed. Error: Command timed out (300s): "cmd /c 'C:\ceph\unittest_fault_injector.exe --
```

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
jenkins test windows

**Review by tchaikov — APPROVED** [review]

**Comment by petrutlucian94** [comment]
jenkins test make check arm64

**Comment by petrutlucian94** [comment]
jenkins test make check arm64

**Pull request merged**